File: 05. CART (Regresion).py
# -*- coding: utf-8 -*-
"""
Created on Thu Aug  5 15:22:38 2021

@author: Admin
"""

# Classification and regression trees

## Usado para regresión, no para clasificación

# =============================================================================
# Libraries
# =============================================================================

import pandas as pd
#import numpy as np
import matplotlib.pyplot as plt
import sklearn.tree as tree

# =============================================================================
# Read data
# =============================================================================

position_salaries = pd.read_csv("data\Position_Salaries.csv")

x = position_salaries.iloc[:, 1:2].values # es mejor que siempre sea matriz
y = position_salaries.iloc[:, 2].values

# =============================================================================
# Ajustar CART
# =============================================================================

regresion = tree.DecisionTreeRegressor(random_state = 0)
regresion.fit(x, y)

# =============================================================================
# Visualizacion
# =============================================================================

plt.scatter(x, y, color = "darkred")
plt.plot(x, regresion.predict(x), color = "darkblue")
plt.title("Modelo CART (Regresion)")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo en $")
plt.show()

# =============================================================================
# Prediccion
# =============================================================================

regresion.predict([[6.5]])

File: 05. CART (Regresion).R
## CART ##

# Uso de CART para regresión, usa la suma de cuadrados


# Libraries ---------------------------------------------------------------

library(tidyverse)
library(rpart)
library(rpart.plot)

# read data ---------------------------------------------------------------

position_salaries <- readr::read_csv("data/Position_Salaries.csv") %>% 
  janitor::clean_names()

position_salaries <- position_salaries %>% 
  dplyr::select(-position)

# ajuste del modelo -------------------------------------------------------

regresion <- position_salaries %>% 
  rpart::rpart(salary ~ ., data = .,
               control = rpart::rpart.control(minsplit = 2))


# visualizacion -----------------------------------------------------------

## Arbol

rpart.plot::prp(regresion)

## Predicción

position_salaries %>% 
  dplyr::mutate(ajustados = predict(regresion)) %>% 
  ggplot2::ggplot(aes(level, y = salary)) + 
  geom_point(color = "darkred") +
  geom_line(aes(y = ajustados), color = "darkblue") +
  labs(title = "CART (Regresion)",
       x = "Nivel del empleado",
       y = "Salario ($)") +
  theme_classic()


# predicción --------------------------------------------------------------

predict(regresion, data.frame(level = 6.5))
